crypto/params.c

```diff
@@ -723,3 +723,392 @@ OSSL_PARAM OSSL_PARAM_construct_octet_ptr(const char *key, void **buf,
+}
+
+int OSSL_PARAM_locate_set_size_t(const OSSL_PARAM *p, const char *key,
+                                 size_t val)
+{
+    return OSSL_PARAM_set_size_t(OSSL_PARAM_locate(p, key), val);
+}
+
+int OSSL_PARAM_locate_get_double(const OSSL_PARAM *p, const char *key,
+                                 double *val)
+{
+    return OSSL_PARAM_get_double(OSSL_PARAM_locate(p, key), val);
+}
+
+int OSSL_PARAM_locate_set_double(const OSSL_PARAM *p, const char *key,
+                                 double val)
+{
+    return OSSL_PARAM_set_double(OSSL_PARAM_locate(p, key), val);
+}
+
+int OSSL_PARAM_locate_get_BN(const OSSL_PARAM *p, const char *key,
+                             BIGNUM **val)
+{
+    return OSSL_PARAM_get_BN(OSSL_PARAM_locate(p, key), val);
+}
+
+int OSSL_PARAM_locate_set_BN(const OSSL_PARAM *p, const char *key,
+                             const BIGNUM *val)
+{
+    return OSSL_PARAM_set_BN(OSSL_PARAM_locate(p, key), val);
+}
+
+int OSSL_PARAM_locate_get_utf8_string(const OSSL_PARAM *p, const char *key,
+                                      char **val, size_t max_len)
+{
+    return OSSL_PARAM_get_utf8_string(OSSL_PARAM_locate(p, key), val, max_len);
+}
+
+int OSSL_PARAM_locate_set_utf8_string(const OSSL_PARAM *p, const char *key,
+                                      const char *val)
+{
+    return OSSL_PARAM_set_utf8_string(OSSL_PARAM_locate(p, key), val);
+}
+
+int OSSL_PARAM_locate_get_octet_string(const OSSL_PARAM *p, const char *key,
+                                       void **val, size_t max_len,
+                                       size_t *used_len)
+{
+    return OSSL_PARAM_get_octet_string(OSSL_PARAM_locate(p, key), val, max_len,
+                                       used_len);
+}
+
+int OSSL_PARAM_locate_set_octet_string(const OSSL_PARAM *p, const char *key,
+                                       const void *val, size_t len)
+{
+    return OSSL_PARAM_set_octet_string(OSSL_PARAM_locate(p, key), val, len);
+}
+
+int OSSL_PARAM_locate_get_utf8_ptr(const OSSL_PARAM *p, const char *key,
+                                   const char **val)
+{
+    return OSSL_PARAM_get_utf8_ptr(OSSL_PARAM_locate(p, key), val);
+}
+
+int OSSL_PARAM_locate_set_utf8_ptr(const OSSL_PARAM *p, const char *key,
+                                   const char *val)
+{
+    return OSSL_PARAM_set_utf8_ptr(OSSL_PARAM_locate(p, key), val);
+}
+
+int OSSL_PARAM_locate_get_octet_ptr(const OSSL_PARAM *p, const char *key,
+                                    const void **val, size_t *used_len)
+{
+    return OSSL_PARAM_get_octet_ptr(OSSL_PARAM_locate(p, key), val, used_len);
+}
+
+int OSSL_PARAM_locate_set_octet_ptr(const OSSL_PARAM *p, const char *key,
+                                    const void *val, size_t used_len)
+{
+    return OSSL_PARAM_set_octet_ptr(OSSL_PARAM_locate(p, key), val, used_len);
+}
+
+int OSSL_PARAM_optional_get_int(const OSSL_PARAM *p, const char *key, int *val)
```

Comment:
The optional "get" forms don't seem very useful. With a "success" return you can't tell whether `key` and `val` have been updated or not. Applications would be better off just using OSSL_PARAM_locate_get_int which returns fail if it wasn't found.

Reply:
Let's say we have a provider that offers RSA, and with which applications (or libcrypto) can set the RSA object parameters *n*, *e* and *d*.

The application (or libcrypto for that matter) wants to create a public key, so tries this:

```
    OSSL_PARAM params[3];
    size_t n = 0;
    unsigned char e_buf[16];
    unsigned char n_buf[512];

    BN_bn2nativepad(rsa_e, e_buf, sizeof(e_buf));
    BN_bn2nativepad(rsa_n, n_buf, sizeof(n_buf));

    params[n++] = OSSL_PARAM_construct_BN("e", rsa_e, sizeof(rsa_e), NULL);
    params[n++] = OSSL_PARAM_construct_BN("n", rsa_n, sizeof(rsa_n), NULL);
    params[n].key = NULL;

    if (!EVP_PKEY_set_params(rsa_pkey, params))
        fprintf(stderr, "ERROR\n");
```

Now, on the provider end, `set_params` is defined like this:

``` C
static int rsa_set_params(void *vrsa, const OSSL_PARAM *params)
{
    PROV_RSA *rsa = vrsa;
    return OSSL_PARAM_locate_get_BN(params, "n", rsa->n)
        && OSSL_PARAM_locate_get_BN(params, "e", rsa->e)
        && OSSL_PARAM_locate_get_BN(params, "d", rsa->d);
}
```

With this construction, creating a public key with parameters fails.

Reply:
Let's say we have a provider that offers RSA, and with which applications (or libcrypto) can set the RSA object parameters *n*, *e* and *d*.

The application (or libcrypto for that matter) wants to create a public key, so tries this:

```
    OSSL_PARAM params[3];
    size_t n = 0;
    unsigned char e_buf[16];
    unsigned char n_buf[512];

    BN_bn2nativepad(rsa_e, e_buf, sizeof(e_buf));
    BN_bn2nativepad(rsa_n, n_buf, sizeof(n_buf));

    params[n++] = OSSL_PARAM_construct_BN("e", rsa_e, sizeof(rsa_e), NULL);
    params[n++] = OSSL_PARAM_construct_BN("n", rsa_n, sizeof(rsa_n), NULL);
    params[n].key = NULL;

    if (!EVP_PKEY_set_params(rsa_pkey, params))
        fprintf(stderr, "ERROR\n");
```

Now, on the provider end, `set_params` is defined like this:

``` C
static int rsa_set_params(void *vrsa, const OSSL_PARAM *params)
{
    PROV_RSA *rsa = vrsa;
    return OSSL_PARAM_locate_get_BN(params, "n", rsa->n)
        && OSSL_PARAM_locate_get_BN(params, "e", rsa->e)
        && OSSL_PARAM_locate_get_BN(params, "d", rsa->d);
}
```

With this construction, creating a public key with parameters fails.

Reply:
Before saying that this won't happen, consider `test/testrsapub.pem`.  How do you imagine that this file will become a default provider PKEY?

Reply:
Right, so "d" truly is optional in this case. So write it like this:

````
static int rsa_set_params(void *vrsa, const OSSL_PARAM *params)
{
    PROV_RSA *rsa = vrsa;

    OSSL_PARAM_locate_get_BN(params, "d", rsa->d);
    return OSSL_PARAM_locate_get_BN(params, "n", rsa->n)
        && OSSL_PARAM_locate_get_BN(params, "e", rsa->e);
}
````

Reply:
And then, another dude just wants to sign with a private key and passes *d* and *n*...

And oh, what if the passed *d* param was invalid?

Reply:
> And oh, what if the passed d param was invalid?

Hmm. Ok.

Reply:
My concern is the fragility of an optional get.

Only supplying _n_ and _e_ but not _d_ in the example above is perfectly reasonable.  At this point, however, it is up to the caller to be careful about not doing anything that might need _d_ and the receiver of these params is unable to verify that an operation makes sense.  This is fragile.

The _d_ value could be anything (from a different key e.g.).  This is bad.

Better, I think, for the receiver to know which values could be extracted.  If only to allow better error detection.

Reply:
An optional get with a default value or an optional get that NULLed values would be more palatable.

Reply:
Maybe it could be left as an exercise to those who write the receiving end to figure out what values have to be grouped together and how to code that properly?

I'm getting worried that this API gets over engineered and tries to solve everything instead of being a set of tools. That comes with a cost.

Reply:
I'd prefer to lose the optional get calls in favour of the receiver figuring it out.

Reply:
The way I see it now, that speaks for this PR in its entirety.

Reply:
Okay, I'll close unless somebody does think this is beneficial.

Reply:
I tend to agree.